rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,223 @@
+---
+feature: pipe-operator
+start-date: 2023-05-23
+author: @piegamesde
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a new "pipe" operator, `|>`, to the Nix language, defined as `f a` = `a |> f`.
+Additionally, elevate `lib.pipe` to a built-in function.
+
+As a reminder, `pipe a [ f g h ]` is defined as `h (g (f a))`.
+
+# Motivation
+[motivation]: #motivation
+
+Creating advanced data processing like transforming a list is a thing commonly done in nixpkgs.
+Yet the language has no support for function concatentation/composition,
+which results in such constructs looking unwieldy and difficult to format well.
+`lib.pipe` may be the most powerful library function with that regard,
+but it is unknown and overlooked by many because it is not easily discoverable:
+Despite its great usefulness, it is currently used in less than 30 files in Nixpkgs
+(`rg '[\. ]pipe .* \['`).
+Additionally, it is not accessible to Nix code outside of nixpkgs,
+and due to Nix's lazy evaluation debugging type errors is really difficult.
+
+Let's have a look at an arbitrarily chosen snippet of Nixpkgs code:
+
+```nix
+defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (lib.concatStringsSep "\n" (lib.mapAttrsToList (key: value: ''
+  // ${value.reason}
+  pref("${key}", ${builtins.toJSON value.value});
+'') defaultPrefs));
+```
+
+It is arguably pretty hard to read and reason about. Even when applying some more whitespace-generous formatting:
+
+```nix
+defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (
+  lib.concatStringsSep "\n" (
+    lib.mapAttrsToList
+    (
+      key: value: ''
+        // ${value.reason}
+        pref("${key}", ${builtins.toJSON value.value});
+      ''
+    )
+    defaultPrefs
+  )
+);
+```
+
+One can observe the following issues:
+
+- If you want to follow the data flow, you must read it from bottom to top,
+  from the inside to the outside (the input here is `defaultPrefs`).
```

Comment:
Conversely, if you want to follow the evaluation order, you must read it in its natural order, from top to bottom.

This negates the point for me.

Note that recent Nix releases (2.13?) also show error traces in evaluation order: the last steps are printed last.

Mini explanation:

```
nix-repl> f = builtins.trace "f" (x: builtins.trace "f body" x) 

nix-repl> g = builtins.trace "g" (y: builtins.trace "g body" y) 

nix-repl> f (g null)
trace: f
trace: f body
trace: g
trace: g body
null
```

Reply:
Most of the time when writing Nix code, I don't really care about the specific order in which Nix evaluates things. I care a lot more about the intention and the order of the steps that process my data.

That traces show in evaluation order is not a contradiction, because traces are mostly used for debugging, which is also one of the rare situations where I do care about evaluation order.

Reply:
Shepherds currently believe this conversation is resolved, outcome: there is value to a `|>` operator, inclined to leave `pipe` in `lib`.